Add python backend support

thakkarparth007 opened the pull request:
**Changes**:
- Modify dockerfile to include bitsandbytes, transformers and latest version of pytorch
- Minor modifications in utils/codegen.py so that same client works with FT and Py-backend
- Minor modifications in launch.sh (no need to name models by GPU)
- Add installation script for adding a new python model (with super simple config_template)
- Modify setup.sh so that it aworks with both FT and Python backend models

I've tested the workflow from top to bottom and it seems to work for me.

**Limitations**:
- Many parameters are unused while generating the output (e.g. beam_width). See `config_template.pbtxt` for unused parameters.
- Doesn't take care of batching multiple requests together.
- Slow performance. FasterTransformer seems to be 6-7x faster than the python backend. But this is still useful if you want to play with different models, or try out a bigger model (e.g., I can fit the 2B model in my tiny 4GB GPU which I can't with FT). Some numbers: for ~20 tokens it takes ~120ms for FT while 800-900ms for Python backend. Most of Python backend's time goes in `model.generate`.
- Didn't test with multi-gpu world (but should work because of `device_map="auto"`)
- Code limited to installing codegen-{350M,2B}-{multi,mono} models. Can trivially allow others, but I'm not sure of the memory requirements of those (it _should_ be O(model_params/1B) + constant GB).
- Didn't modify README yet.
 
**Caveats**:
- Model directory style is slightly different because we don't need a separate copy of the model based on the number of available GPUs. Accelerate takes care of that for us.
- Haven't tested with models other than codegen models. The code should work for all decoder-only models IMO, and will probably need to be tweaked a bit while working with Encoder-Decoder models.

Signed-off-by: Parth Thakkar <thakkarparth007@gmail.com>

moyix commented:
Oh this is wonderful! This was one of my big wishlist items :D I will try to take this for a test drive later this week and hopefully get it merged shortly after #73 goes in!

fdegier commented on Dockerfile:
```diff
@@ -0,0 +1,5 @@
+
+# Install dependencies: torch
+RUN pip3 install -U torch --extra-index-url https://download.pytorch.org/whl/cu116
+RUN pip3 install -U transformers bitsandbytes accelerate
```
`moyix/triton_with_ft:22.09` comes from this repo https://github.com/moyix/fastertransformer_backend/blob/main/docker/Dockerfile

Seems to me more logical to add the dependencies there?

fdegier reviewed (changes requested):
I've made a pr for you that resolved the merge conflicts etc. https://github.com/thakkarparth007/fauxpilot/pull/1

thakkarparth007 commented on copilot_proxy/models.py:
```diff
@@ -4,7 +4,7 @@
 
 class OpenAIinput(BaseModel):
-    model: str
+    model: str = "fastertransformer|py-model"
```
Oh it was meant to signal users that both fastertransformer and py-model are valid values. I wasn't sure how else to convey that, so ended up using this unusual format. I guess a better way to convey that would be to just mark it in the README/docs and let the user who wants to use python backend set that on their own.

Or maybe there's a way to annotate this information such that it shows up in Swagger UI. I'll try to see if there's such an option or fallback to the above option.

thakkarparth007 commented on docker-compose.yaml:
```diff
@@ -1,11 +1,14 @@
-    image: moyix/triton_with_ft:22.09
+    build:
+      context: .
+      dockerfile: Dockerfile
```
Sounds good to me. I can make a PR to that repo instead

fdegier commented on docker-compose.yaml:
```diff
@@ -1,11 +1,14 @@
     shm_size: '2gb'
     volumes:
       - ${MODEL_DIR}:/model
+      - ${HF_CACHE_DIR}:/root/.cache/huggingface
```
If no `HF_CACHE_DIR` is set this breaks the deployment. Maybe we should default to true?

thakkarparth007 replied:
The current version sets HF_CACHE_DIR to `/tmp/hf_cache`. The reason I didn't set it to default to true was because docker messes up the file permissions by setting many of them to root (unless rootless docker is used). So, the cache is shared only if the user knows what they're doing.

I could default it to true, and warn the user about the permission issues. Not sure which is a good option.

@fdegier, @moyix thoughts?

fdegier replied:
@thakkarparth007 if a user does not use the cache, the volume is set to empty, which can not be mounted and causes the `docker compose up` to fail, hence why I suggested to default to true but I understand the permission issues you mentioned. 

I think adding  something like `HF_DATASETS_CACHE="fauxpilot/.hf-cache"` and removing the option to cache, will always cache and not mess up permissions?

thakkarparth007 replied:
Whoops, I just saw your comment @fdegier 

So currently if you notice in the setup.sh (https://github.com/moyix/fauxpilot/pull/86/files#diff-4209d788ad32c40cbda3c66b3de47eefb929308ca703bb77a6382625986add17R148) then you'll see that HF_CACHE_DIR is being set to `/tmp/hf_cache` if the user doesn't want to share hf_cache.

But yes, perhaps it'll be better to store the cache in the fauxpilot directory itself.

thakkarparth007 commented:
@fdegier thanks for your [PR](https://github.com/thakkarparth007/fauxpilot/pull/1) that merges python_backend code on top of  #73 

I made a few small changes on top of those (except your last commit since my changes were already in progress). Specifically, fixed some issues in the setup.sh file and added a test_setup.py script which performs an E2E test of the python backend functionality.

@moyix this test script builds towards your CI feature from the wishlist. I've run the test locally (just need to invoke the script using `pytest`). The script currently doesn't need GPU so it can be integrated in a free tier CI tool.

With a few modifications, it can be used for checking fastertransformer backend as well. I don't know if fastertransformer backend can work without GPU though.

thakkarparth007 commented:
Update: Fixed the test.

For running the test, you can just run `pytest tests` from the root directory.
